pkg/etcd/client.go

```diff
@@ -165,7 +177,62 @@ func (c *Client) TimeToLive(ctx context.Context, lease clientv3.LeaseID, opts ..
 
 // Watch delegates request to clientv3.Watcher.Watch
 func (c *Client) Watch(ctx context.Context, key string, opts ...clientv3.OpOption) clientv3.WatchChan {
-	return c.cli.Watch(ctx, key, opts...)
+	watchCh := make(chan clientv3.WatchResponse, etcdWatchChBufferSize)
+	go c.WatchWithChan(ctx, watchCh, key, opts...)
+	return watchCh
+}
+
+// WatchWithChan maintains a watchCh and send all msg from the watchCh to outCh
+func (c *Client) WatchWithChan(ctx context.Context, outCh chan clientv3.WatchResponse, key string, opts ...clientv3.OpOption) {
+	defer log.Info("WatchWithChan exited")
+	var lastRevision int64
+	watchCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	watchCh := c.cli.Watch(watchCtx, key, opts...)
+
+	watchTicker := c.clock.Ticker(etcdWatchChTimeoutDuration)
+	defer watchTicker.Stop()
+	requestTicker := c.clock.Ticker(etcdRequestProgressDuration)
+	defer requestTicker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			cancel()
+			return
+		case response := <-watchCh:
+			start := c.clock.Now()
+			if response.Err() == nil {
+				lastRevision = response.Header.Revision
+			}
+
+		Loop:
+			for {
+				select {
+				case <-ctx.Done():
+					cancel()
+					return
+				case outCh <- response: // it may blocking here
+					break Loop
+				case <-watchTicker.C:
+					log.Warn("etcd client outCh blocking too long, the etcdWorker may be stuck", zap.Duration("duration", c.clock.Since(start)))
+				}
+			}
+
+			watchTicker.Reset(etcdWatchChTimeoutDuration)
```

Comment:
We don't need to reset the ticker. But we need to track the time when we last received a message from `watchCh`, so that "spurious wake-ups" do not cause false positives.